src/common/async.cpp

```diff
@@ -4,156 +4,172 @@
-                        // not own the lock. For this reason it's important
-                        // to only modify state->active in the current thread,
-                        // i.e. in the background thread (i.e this function)
-                        //
-                        state->active.erase(pair.first);
-                        state->changed = true;
-                        debug("async: test stopped");
-                        if (pair.second) {
-                            pair.second(pair.first);
-                        }
+                NetTest *ptr = pair.first;
+                debug("async: active: %lld", ptr->identifier());
+                INSERT(state->active, ptr, pair.second);
+                // Note: it should be safe to capture `state` by reference
+                // here because the destructor should not be able to destroy
+                // the background thread until we detach it and break.
+                ptr->begin([&state, ptr]() {
+                    ptr->end([&state, ptr]() {
+                        // Safe to lock because loop_once() is unlocked
+                        LOCKED({
+                            debug("async: need to lock to move test object");
+                            NetTestVar var;
+                            GET(var, state->active, ptr);
+                            INSERT(state->completed, ptr, var);
+                            state->active.erase(ptr);
+                            state->changed = true;
+                            debug("async: completed: %lld", var->identifier());
+                            debug("async: done with briefly locking");
+                        })
                     });
                 });
             }
             state->ready.clear();
-        )
+        })
 
         debug("async: loop thread unlocked");
-
+        state->changed = false;
         while (!state->changed) {
-            //state->poller->loop_once();
             loop_once();
         }
-        state->changed = false;
-
         debug("async: bottom of loop thread");
     }
-    debug("async: thread detached");
-    state->thread.detach();
-    state->thread_running = false;
-    if (state->hook_empty) {
-        state->hook_empty();
-    }
     debug("async: exiting from thread");
 }
 
-Async::Async(SharedPointer<Poller> poller) {
-    state.reset(new AsyncState());
-    state->poller = poller;
-}
+Async::Async() { state.reset(new AsyncState()); }
 
-void Async::run_test(SharedPointer<NetTest> test,
-  std::function<void(SharedPointer<NetTest>)> fn) {
-    LOCKED(
-        debug("async: test inserted");
-        auto ret = state->ready.insert(std::make_pair(test, fn));
-        if (!ret.second) {
-            throw std::runtime_error("async: element already inserted");
-        }
+void Async::run_test(NetTestVar test, std::function<void(NetTestVar)> fn) {
+    LOCKED({
+        NetTest *ptr = test.operator->(); // get() does not check for NULL
```

Comment:
Actually since 64c9992 `SharedPointer::get()` checks for `nullptr` so we can modify the above to use `get()`.